discovery-provider/src/tasks/entity_manager/delegation.py

```diff
@@ -0,0 +1,156 @@
+import json
+from typing import Optional, TypedDict, Union, cast
+
+from src.models.delegates.delegation import Delegation
+from src.tasks.entity_manager.utils import (
+    Action,
+    EntitySpecialFetchType,
+    EntityType,
+    ManageEntityParameters,
+)
+from src.utils.indexing_errors import EntityMissingRequiredFieldError
+from src.utils.model_nullable_validator import all_required_fields_present
+from src.utils.structured_logger import StructuredLogger
+
+logger = StructuredLogger(__name__)
+
+
+class CreateDelegationMetadata(TypedDict):
+    shared_address: Union[str, None]
+    delegate_address: Union[str, None]
+
+
+def get_create_delegation_metadata_from_raw(
+    raw_metadata: Optional[str],
+) -> Optional[CreateDelegationMetadata]:
+    metadata: CreateDelegationMetadata = {
+        "shared_address": None,
+        "delegate_address": None,
+    }
+    if raw_metadata:
+        try:
+            json_metadata = json.loads(raw_metadata)
+            raw_shared_address = json_metadata.get("shared_address", None)
+            if raw_shared_address:
+                metadata["shared_address"] = raw_shared_address.lower()
+            else:
+                metadata["shared_address"] = None
+            raw_delegate_address = json_metadata.get("delegate_address", None)
+            if raw_delegate_address:
+                metadata["delegate_address"] = raw_delegate_address.lower()
+            else:
+                metadata["delegate_address"] = None
+            return metadata
+        except Exception as e:
+            logger.error(
+                f"entity_manager | delegation.py | Unable to parse delegate metadata while indexing: {e}"
+            )
+            return None
+    return metadata
+
+
+def validate_delegation_tx(
+    params: ManageEntityParameters, metadata: CreateDelegationMetadata
+):
+    user_id = params.user_id
+
+    if params.entity_type != EntityType.DELEGATION:
+        raise Exception(
+            f"Invalid Delegation transaction, wrong entity type {params.entity_type}"
+        )
+
+    if params.action == Action.CREATE:
+        if not metadata["shared_address"]:
+            raise Exception(
+                "Invalid Delegation transaction, shared address is required and was not provided"
+            )
+        if not metadata["delegate_address"]:
+            raise Exception(
+                "Invalid Delegation transaction, delegate address is required and was not provided"
+            )
+        if (
+            metadata["delegate_address"].lower()
```

Comment:
delegate_address is an optional but i think it's ok to just expect non null values for CreateDelegationMetadata